### PR TITLE
Add basic sentry integration

### PR DESCRIPTION
**Pull request by ramon-juicelabs**
This includes moving the logger to use zap allowing some nice hooks for sentry and stack trace functionality in general for logging

### CODE REVIEW

File: cmd/agent/main.go [hunk 1]
```diff
@@ -25,7 +27,20 @@ var (
...
 func main() {
-	appmain.Run("Juice Agent", build.Version, func(group task.Group) error {
+	name := "Juice Agent"
+	config := appmain.Config{
+		Name:    name,
+		Version: build.Version,
+
+		SentryConfig: sentry.ClientOptions{
+			Release:          fmt.Sprintf("%s@%s", name, build.Version),
+			Dsn:              "https://4bd058ff6346e984d53edcba7c22ea63@o4505739073486848.ingest.sentry.io/4505763525427200",
```
Comment (dave-juicelabs) [inline cmd/agent/main.go:37]: Do we want this hardcoded in the repo? Maybe an env variable during development?

File: cmd/controller/main.go [hunk 2]
```diff
@@ -89,7 +90,20 @@ func openStorage(ctx context.Context) (storage.Storage, error) {
...
 func main() {
-	appmain.Run("Juice Controller", build.Version, func(group task.Group) error {
+	name := "Juice Controller"
+	config := appmain.Config{
+		Name:    name,
+		Version: build.Version,
+
+		SentryConfig: sentry.ClientOptions{
+			Release:          fmt.Sprintf("%s@%s", name, build.Version),
+			Dsn:              "https://bd72f965550ed6269cd8c3604aa0c769@o4505739073486848.ingest.sentry.io/4505763982147584",
```
Comment (dave-juicelabs) [inline cmd/controller/main.go:100]: Same as above

File: cmd/juicify/main.go [hunk 3]
```diff
@@ -4,14 +4,29 @@
...
-	appmain.Run("juicify", build.Version, func(group task.Group) error {
+	name := "juicify"
+
+	config := appmain.Config{
+		Name:    name,
+		Version: build.Version,
+
+		SentryConfig: sentry.ClientOptions{
+			Release: fmt.Sprintf("%s@%s", name, build.Version),
+			Dsn:     "https://fb7e2006c23d07f7b3ba78067164eba4@o4505739073486848.ingest.sentry.io/4505779685818368",
```
Comment (dave-juicelabs) [inline cmd/juicify/main.go:25]: Same as above

File: pkg/appmain/appmain.go [hunk 4]
```diff
@@ -28,27 +36,37 @@ var (
...
 	}
 
-	err := logger.Configure()
-	if err == nil {
-		logger.Info(name, ", v", version)
+	if err = sentry.Initialize(config.SentryConfig); err == nil {
+
+		err = logger.Configure()
+		if err == nil {
+			defer sentry.Close()
```
Comment (dave-juicelabs) [inline pkg/appmain/appmain.go:53]: This should be moved outside this if statement

File: pkg/appmain/appmain.go [hunk 5]
```diff
@@ -28,27 +36,37 @@ var (
...
+			ctx, _ := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
 
-		taskManager := task.NewTaskManager(ctx)
-		taskManager.GoFn("AppMain", logic)
-		err = taskManager.Wait()
+			taskManager := task.NewTaskManager(ctx)
+			taskManager.GoFn("AppMain", logic)
+			err = taskManager.Wait()
+			if err != nil {
+				logger.Error(err)
```
Comment (dave-juicelabs) [inline pkg/appmain/appmain.go:67]: Add os.Exit(ExitFailure) here. No reason to log the error here just for it to logged again down below. Instead, logging to sentry and then exiting would be good.

**Review by dave-juicelabs — DISMISSED** [review]

File: pkg/appmain/appmain.go (hunk 5 above)
Comment (ramon-juicelabs) [reply pkg/appmain/appmain.go:66]: os.Exit unfortunately doesn't call the deferred functions so anything through sentry/logger is lost if it hasn't been sent. I might adjust a few things so the error is returned from appmain.Run(...) and let the caller manage exiting etc

File: cmd/agent/main.go (hunk 1 above)
Comment (ykagan) [reply cmd/agent/main.go:37]: you could put this in `.env` - it should get loaded on startup into env

File: pkg/logger/logger.go [hunk 6]
```diff
@@ -6,199 +6,122 @@ package logger
...
-	LevelDebug
-	LevelTrace
+	"go.uber.org/zap"
 )
 
 var (
 	quiet       = flag.Bool("quiet", false, "Disables all logging output")
 	logLevelArg = flag.String("log-level", "info", "Sets the maximum level of output [Fatal, Error, Warning, Info (Default), Debug, Trace]")
 	logFile     = flag.String("log-file", "", "")
+	logFormat   = flag.String("log-format", "juice", "Set the format of the logging [juice, console, json]")
```
Comment (dave-juicelabs) [inline pkg/logger/logger.go:18]: What is the difference between juice and console?

File: pkg/appmain/appmain.go (hunk 5 above)
Comment (ramon-juicelabs) [reply pkg/appmain/appmain.go:66]: Ah apologies @dave-juicelabs I thought you were referring to a os.Exit I had in there previously. I haven't actually made the change I said I was going to do just yet. But the logic here is that if the logger fails to init the error will be printed to stderr, while logging will goto stdout or a file.

File: pkg/logger/logger.go (hunk 6 above)
Comment (ramon-juicelabs) [reply pkg/logger/logger.go:18]: Console is zap's development display - probably should just gate it to juice or json. The console one is pretty nice but doesn't display a single line for anything > INFO or something as it'll display a call stack in those situations

File: Build-All.ps1 [hunk 7]
```diff
@@ -25,15 +25,33 @@ if ($IsWindows)
...
+        agent       = $env:JUICE_AGENT_SENTRY_DSN;
+        controller  = $env:JUICE_CONTROLLER_SENTRY_DSN;
+        juicify     = $env:JUICE_JUICIFY_SENTRY_DSN;
+    }
+
+    $dsn = $SentryDsn[$Component]
+    
+    return @(
+        "-X github.com/Juice-Labs/Juice-Labs/cmd/internal/build.Version=$Version";
+        "-X github.com/Juice-Labs/Juice-Labs/pkg/sentry.SentryDsn=$dsn";
```
[the remark this answers is not in the document]
Comment (ramon-juicelabs) [reply Build-All.ps1:40]: Not putting in a DSN is OK as it'll just not initialize Sentry. The tricky problem I have found is where does this information go - I think I've found a solution, see the native PR as that is the source of truth for building local/dev builds. The current state in this repo is taking care of via Github variables.

File: pkg/appmain/appmain.go [hunk 8]
```diff
@@ -28,31 +36,40 @@ var (
...
+	ctx, _ := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
 
+	taskManager := task.NewTaskManager(ctx)
+	taskManager.GoFn("AppMain", logic)
+	err = taskManager.Wait()
 	if err != nil {
-		fmt.Fprintln(os.Stderr, err)
-		os.Exit(ExitFailure)
+		logger.Error(err)
 	}
```
Comment (dave-juicelabs) [inline pkg/appmain/appmain.go:74]: Still need the os.Exit(ExitFailure) here so that the error code is correct when the process exits.

**Review by dave-juicelabs — CHANGES_REQUESTED** [review]

**Review by dave-juicelabs — APPROVED** [review]

**Pull request merged**